sys/log/full/src/log_fcb2.c

```diff
@@ -0,0 +1,890 @@
+    }
+
+    /* Set log watermark to end of this element */
+    if (ueh.ue_index >= log_off->lo_index) {
+        fl->fl_watermark_sec = loc->fe_sector;
+        fl->fl_watermark_off = loc->fe_data_off + loc->fe_data_len;
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+static int
+log_fcb2_set_watermark(struct log *log, uint32_t index)
+{
+    int rc;
+    struct log_offset log_offset;
+    struct fcb_log *fl;
+    struct fcb *fcb;
+
+    fl = (struct fcb_log *)log->l_arg;
+    fcb = &fl->fl_fcb;
+
+    log_offset.lo_arg = NULL;
+    log_offset.lo_ts = 0;
+    log_offset.lo_index = index;
+    log_offset.lo_data_len = 0;
+
+    /* Find where to start the walk, and set watermark accordingly */
+    rc = log_fcb2_walk(log, log_fcb2_new_watermark_index, &log_offset);
+    if (rc != 0) {
+        goto done;
+    }
+
+    /* If there are no entries to read and the watermark has not been set */
+    if (fl->fl_watermark_off == 0xffffffff) {
+        fl->fl_watermark_sec = fcb->f_oldest_sec;
+        fl->fl_watermark_off = 0;
+    }
+
+    return (0);
+done:
+    return (rc);
+}
+#endif
+
+/**
+ * Copies one log entry from source fcb to destination fcb
+ * @param src_fcb, dst_fcb
+ * @return 0 on success; non-zero on error
+ */
+static int
+log_fcb2_copy_entry(struct log *log, struct fcb_entry *entry,
+                    struct fcb *dst_fcb)
+{
+    struct log_entry_hdr ueh;
+    char data[LOG_PRINTF_MAX_ENTRY_LEN + LOG_BASE_ENTRY_HDR_SIZE +
+              LOG_IMG_HASHLEN];
+    uint16_t hdr_len = 0;
+    int dlen;
+    int rc;
+    struct fcb *fcb_tmp;
+
+    rc = log_fcb2_read(log, entry, &ueh, 0, LOG_BASE_ENTRY_HDR_SIZE);
+    if (rc != LOG_BASE_ENTRY_HDR_SIZE) {
+        goto err;
+    }
+
+    hdr_len = log_hdr_len(&ueh);
+
+    dlen = min(entry->fe_data_len, LOG_PRINTF_MAX_ENTRY_LEN + hdr_len);
+
+    rc = log_fcb2_read(log, entry, data, 0, dlen);
+    if (rc < 0) {
+        goto err;
+    }
+
+    /* Changing the fcb to be logged to be dst fcb */
+    fcb_tmp = log->l_arg;
+    log->l_arg = dst_fcb;
+    rc = log_fcb2_append(log, data, dlen);
+    log->l_arg = fcb_tmp;
+    if (rc) {
+        goto err;
+    }
+
+err:
+    return (rc);
+}
+
+/**
+ * Copies log entries from source fcb to destination fcb
+ * @param src_fcb, dst_fcb, element to start copying from
+ * @return 0 on success; non-zero on error
+ */
+static int
+log_fcb2_copy(struct log *log, struct fcb *src_fcb, struct fcb *dst_fcb,
+              struct fcb_entry *from)
+{
+    struct fcb_entry entry;
+    int rc;
+
+    rc = 0;
+
+    entry = *from;
+    do {
+        rc = log_fcb2_copy_entry(log, &entry, dst_fcb);
+        if (rc) {
+            break;
+        }
+        rc = fcb_getnext(src_fcb, &entry);
+        if (rc == FCB_ERR_NOVAR) {
+            rc = 0;
+            break;
+        }
+    } while (rc == 0);
+
+    return (rc);
+}
+
+/**
+ * Flushes the log while restoring specified number of entries
+ * using image scratch
+ * @param src_fcb, dst_fcb
```

Comment:
params do not match comment, which is also enigmatic 'specified number of entries'.
It was copied from log_fcb.c and that place could also be updated, but since it was not part of this change I did not put comment there.

Reply:
Yup, the comments don't appear to match the code.